src/Features/CSharp/Portable/ConvertLinq/CSharpConvertForEachToLinqQueryProvider.cs

```diff
@@ -0,0 +1,585 @@
+            private ForEachStatementSyntax _forEachStatement;
+            private ImmutableArray<SyntaxNode> _convertingNodes;
+            private YieldStatementSyntax _yieldStatement;
+            private SyntaxNode _nodeToDelete;
+
+            public YieldReturnConverter(
+                ForEachStatementSyntax forEachStatement,
+                ImmutableArray<SyntaxNode> convertingNodes,
+                YieldStatementSyntax yieldStatement,
+                SyntaxNode nodeToDelete = null)
+            {
+                _forEachStatement = forEachStatement;
+                _convertingNodes = convertingNodes;
+                _yieldStatement = yieldStatement;
+                _nodeToDelete = nodeToDelete;
+            }
+
+            public void Convert(SyntaxEditor editor, SemanticModel semanticModel, CancellationToken cancellationToken)
+            {
+                var queryExpression = CreateQueryExpression(_forEachStatement, _convertingNodes, _yieldStatement.Expression);
+                editor.ReplaceNode(_forEachStatement, SyntaxFactory.ReturnStatement(queryExpression).WithAdditionalAnnotations(Formatter.Annotation));
+                // Delete the yield break just after the loop.
+                if (_nodeToDelete != null)
+                {
+                    editor.RemoveNode(_nodeToDelete);
+                }
+            }
+        }
+
+        private sealed class DefaultConverter : AbstractConverter, IConverter
+        {
+            private ForEachStatementSyntax _forEachStatement;
+            private ImmutableArray<SyntaxNode> _convertingNodes;
+            private ImmutableArray<SyntaxToken> _identifiers;
+            private ImmutableArray<StatementSyntax> _statements;
+
+            public DefaultConverter(
+                ForEachStatementSyntax forEachStatement,
+                ImmutableArray<SyntaxNode> convertingNodes,
+                ImmutableArray<SyntaxToken> identifiers,
+                ImmutableArray<StatementSyntax> statements)
+            {
+                _forEachStatement = forEachStatement;
+                _convertingNodes = convertingNodes;
+                _identifiers = identifiers;
+                _statements = statements;
+            }
+
+            public void Convert(SyntaxEditor editor, SemanticModel semanticModel, CancellationToken cancellationToken)
+            {
+                // Filter out identifiers which are not used in statements.
+                var symbolNames = new HashSet<string>(_statements.SelectMany(statement => semanticModel.AnalyzeDataFlow(statement).ReadInside).Select(symbol => symbol.Name));
+                var identifiersUsedInStatements = _identifiers.Where(identifier => symbolNames.Contains(identifier.ValueText));
+
+                // Wrap statements with a block.
+                var block = WrapWithBlockIfNecessary(_statements.Select(statement => statement.WithoutTrivia().WithTrailingTrivia(SyntaxFactory.ElasticEndOfLine(Environment.NewLine))));
```

Comment:
this comment isn't helpful.  The code already says "WrapWithBlockIfNecessary".  A comment should help explain when/why this happens (or doesn't).